### PR TITLE
Deduplicate category boards

### LINKED ISSUE #4593: Bug: Migration is not working (MM 7.7->7.8)
I'm running MM in a container, and the post-upgrade migration of focalboard fails after upgrading to 7.8.
## Steps to reproduce the behavior

1. Upgrade MM from 7.7 to 7.8
2. Boards are not working / an error is displayed in System Console / Plugin Management

## Expected behavior

Migration should work.

## Edition and Platform

 - Edition: Mattermost Boards (plugin) in a containerized MM serveer
 - Version: 7.8.2

## Additional context (optional)

Error shown in Plugin Management:

> This plugin failed to start. cannot activate plugin: error initializing the DB: driver: postgres, message: failed to execute migration, command: executing_query, originalError: pq: could not create unique index "unique_user_category_board", query: DO $$ BEGIN IF NOT EXISTS ( SELECT * FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS WHERE constraint_schema = 'public' AND constraint_name = 'unique_user_category_board' AND constraint_type = 'UNIQUE' AND table_name = 'focalboard_category_boards' ) THEN ALTER TABLE public.focalboard_category_boards ADD CONSTRAINT unique_user_category_board UNIQUE(user_id, board_id); END IF; END; $$ LANGUAGE plpgsql;

From `focalboard_errors.log` in the container:
```log
error [2023-02-21 10:25:39.915 +01:00] Table creation / migration failed error="driver: postgres, message: failed to execute migration, command: executing_query, originalError: pq: could not create unique index "unique_user_category_board", query:



                DO
                $$
                BEGIN
                IF NOT EXISTS (
                        SELECT * FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS
                                WHERE constraint_schema = 'public'
                                AND constraint_name = 'unique_user_category_board'
                                AND constraint_type = 'UNIQUE'
                                AND table_name = 'focalboard_category_boards'
                ) THEN
                        ALTER TABLE public.focalboard_category_boards ADD CONSTRAINT unique_user_category_board UNIQUE(user_id, board_id);
                END IF;
                END;
                $$
                LANGUAGE plpgsql;




"
  github.com/mattermost/focalboard/server/services/store/sqlstore.New
      github.com/mattermost/focalboard/server/services/store/sqlstore/sqlstore.go:75
  github.com/mattermost/focalboard/mattermost-plugin/server/boards.NewBoardsApp
      github.com/mattermost/focalboard/mattermost-plugin/server/boards/boardsapp.go:91
  main.(*Plugin).OnActivate
      github.com/mattermost/focalboard/mattermost-plugin/server/plugin.go:51
  github.com/mattermost/mattermost-server/v6/plugin.(*hooksRPCServer).OnActivate
      github.com/mattermost/mattermost-server/v6/plugin/client_rpc.go:316
  reflect.Value.call
      reflect/value.go:556
  reflect.Value.Call
      reflect/value.go:339
  net/rpc.(*service).call
      net/rpc/server.go:381
  runtime.goexit
      runtime/asm_amd64.s:1571

error [2023-02-21 10:25:40.129 +01:00] Table creation / migration failed error="driver: postgres, message: failed to execute migration, command: executing_query, originalError: pq: could not create unique index "unique_user_category_board", query:



                DO
                $$
                BEGIN
                IF NOT EXISTS (
                        SELECT * FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS
                                WHERE constraint_schema = 'public'
                                AND constraint_name = 'unique_user_category_board'
                                AND constraint_type = 'UNIQUE'
                                AND table_name = 'focalboard_category_boards'
                ) THEN
                        ALTER TABLE public.focalboard_category_boards ADD CONSTRAINT unique_user_category_board UNIQUE(user_id, board_id);
                END IF;
                END;
                $$
                LANGUAGE plpgsql;




"
  github.com/mattermost/focalboard/server/services/store/sqlstore.New
      github.com/mattermost/focalboard/server/services/store/sqlstore/sqlstore.go:75
  github.com/mattermost/focalboard/mattermost-plugin/server/boards.NewBoardsApp
      github.com/mattermost/focalboard/mattermost-plugin/server/boards/boardsapp.go:91
  main.(*Plugin).OnActivate
      github.com/mattermost/focalboard/mattermost-plugin/server/plugin.go:51
  github.com/mattermost/mattermost-server/v6/plugin.(*hooksRPCServer).OnActivate
      github.com/mattermost/mattermost-server/v6/plugin/client_rpc.go:316
  reflect.Value.call
      reflect/value.go:556
  reflect.Value.Call
      reflect/value.go:339
  net/rpc.(*service).call
      net/rpc/server.go:381
  runtime.goexit
      runtime/asm_amd64.s:1571

error [2023-02-21 10:25:40.129 +01:00] Table creation / migration failed error="driver: postgres, message: failed to execute migration, command: executing_query, originalError: pq: could not create unique index "unique_user_category_board", query:



                DO
                $$
                BEGIN
                IF NOT EXISTS (
                        SELECT * FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS
                                WHERE constraint_schema = 'public'
                                AND constraint_name = 'unique_user_category_board'
                                AND constraint_type = 'UNIQUE'
                                AND table_name = 'focalboard_category_boards'
                ) THEN
                        ALTER TABLE public.focalboard_category_boards ADD CONSTRAINT unique_user_category_board UNIQUE(user_id, board_id);
                END IF;
                END;
                $$
                LANGUAGE plpgsql;




"
  github.com/mattermost/focalboard/server/services/store/sqlstore.New
      github.com/mattermost/focalboard/server/services/store/sqlstore/sqlstore.go:75
  github.com/mattermost/focalboard/mattermost-plugin/server/boards.NewBoardsApp
      github.com/mattermost/focalboard/mattermost-plugin/server/boards/boardsapp.go:91
  main.(*Plugin).OnActivate
      github.com/mattermost/focalboard/mattermost-plugin/server/plugin.go:51
  github.com/mattermost/mattermost-server/v6/plugin.(*hooksRPCServer).OnActivate
      github.com/mattermost/mattermost-server/v6/plugin/client_rpc.go:316
  reflect.Value.call
      reflect/value.go:556
  reflect.Value.Call
      reflect/value.go:339
  net/rpc.(*service).call
      net/rpc/server.go:381
  runtime.goexit
      runtime/asm_amd64.s:1571
```

### PR DESCRIPTION
**Pull request by harshilsharma63**
#### Summary
This PR adds de-duplicate data check and fix to make sure migration number 36 goes in smoothly. Previously we had this check and fix in changelog as important pre-upgrade step but this PR moves that into the product migration itself.

This migration had to be added as a data migration instead of a standard SQL migration because it prepares an existing, already released migration (number 36) to run correctly in case user has duplicate data.

#### Ticket Link
Fixes #4593

### CODE REVIEW

File: server/services/store/sqlstore/data_migrations.go [hunk 1]
```diff
@@ -790,3 +791,102 @@ func (s *SQLStore) getCollationAndCharset(tableName string) (string, string, err
...
+
+func (s *SQLStore) doesDuplicateCategoryBoardsExist() (bool, error) {
+	subQuery := s.getQueryBuilder(s.db).
+		Select("user_id", "board_id", "count(*) AS count").
+		From(s.tablePrefix+"category_boards").
+		GroupBy("user_id", "board_id").
+		Having("count(*) > 1")
+
+	query := s.getQueryBuilder(s.db).
+		Select("COUNT(*)").
```
Comment (wiggin77) [inline server/services/store/sqlstore/data_migrations.go:856]: We would like to get away from using count(*).  See https://community.mattermost.com/core/pl/wqnztqiqotbx9q5mfkonchreco

**Review by sbishel — APPROVED** [review]

**Comment by sbishel** [comment]
@mgdelacroix, @harshilsharma63 - I would like to get this in for `release-7.9` to reduce issues. I need to do a build tomorrow in order for this to happen. But don't want to rush if its not ready or even a little dangerous.

File: server/services/store/sqlstore/data_migrations.go (hunk 1 above)
Comment (harshilsharma63) [reply server/services/store/sqlstore/data_migrations.go:856]: Nice catch @wiggin77 ! I'll update this.

Comment (harshilsharma63) [reply server/services/store/sqlstore/data_migrations.go]: @wiggin77 I've fixed this anyways but I believe the issue you're talking about is with `select *`, not with `select count(*)`.

**Review by mgdelacroix — APPROVED** [review]
LGTM :+1:

**Pull request merged**